### PR TITLE
Consolidate PreFilter and Filter reason messages for scheduler

### LINKED ISSUE #110783: kube-scheduler: improve FailedScheduling message for PreFilter plugin
### What would you like to be added?

**Problem**
`kube-scheduler` assigns the same status to all nodes when a `PreFilter` plugin returns error.https://github.com/kubernetes/kubernetes/blob/11686e138694a67452ce9fe4244ba96982cc545a/pkg/scheduler/schedule_one.go#L380

 The resulted `FailedScheduling` event message can be confusing. For example, when `InterPodAffinity` plugin's `PreFilter` fails to list `NodeInfos`, the scheduling failure event will look like

```
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  31h   default-scheduler  0/134 nodes are available: 134 failed to list NodeInfos ....
```
 
**Proposal** 
Instead of setting the same status/error for all nodes as 
```
        for _, n := range allNodes {
            diagnosis.NodeToStatusMap[n.Node().Name] = s
        }
```

Only set the status once for `PreFilter` error? ` Node().Name `can be a dummy one such as "". 
 ``` 
          diagnosis.NodeToStatusMap[""] = s
```
A prefix string "PreFilter" can also be added to the beginning of the error message. The new `FailedScheduling` event will look like

```
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  31h   default-scheduler  0/134 nodes are available: 1 PreFilter failed to list NodeInfos ....
```

### Why is this needed?

The current `FailedScheduling` event message can be confusing when a `PreFilter` plugin returns an error. By improving the  error message for `PreFilter` plugin,  `FailedScheduling` event will be easier to read and understand.

### PR DESCRIPTION
**Pull request by yuanchen8911**
#### What type of PR is this?

/kind cleanup
#### What this PR does / why we need it:

Add the error messages from `PreFilter` plugin to `FitError.Diagnosis` to make the reasons in `FailedScheduling` more readable and understandable.  The updated FailedScheduling reason message is formatted as the following. 

`0/N nodes are available: <prefilter failure reason>. <filter failure reasons>. <post failure message>.`

An example is

`0/10 nodes are available: Node(s) failed Prefilter plugin A. 2 Node(s) failed Filter plugin B, 8 Node(s) failed Filter plugin C. Error running PostFilter plugin D.`

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/110783

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/types.go [hunk 1]
```diff
@@ -215,6 +215,8 @@ type WeightedAffinityTerm struct {
 type Diagnosis struct {
 	NodeToStatusMap      NodeToStatusMap
 	UnschedulablePlugins sets.String
+	// PreFilterMsg records the messages returned from PreFilterPlugins.
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:218]:
```suggestion
	// PreFilterMsg records the messages returned from PreFilter plugins.
```

(Could you also update PostFilterPlugin to PostFilter plugins?)

File: pkg/scheduler/framework/types.go [hunk 2]
```diff
@@ -238,6 +240,12 @@ func (f *FitError) Error() string {
 		}
 	}
 
+	reasonMsg := fmt.Sprintf(NoNodeAvailableMsg, f.NumAllNodes)
+	// Add the messages from PreFilter to reasonMsg.
+	preFilterMsg := f.Diagnosis.PreFilterMsg
+	if preFilterMsg != "" {
+		reasonMsg += preFilterMsg + " "
+	}
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/types.go:248]: seem we miss `:` here? as I suppose the eventual message would be like:

```
0/X nodes are available: <PreFilterMsg> <FilterMsg> <PostFilterMsg>
```

Given PreFilterMsg and PostFilterMsg can be empty, so we can start the message as `0/X nodes are available:`, and then append "the empty space plus <Msg>" as a chunk afterwards.

Comment (Huang-Wei) [reply pkg/scheduler/framework/types.go:248]: PS1: can we add tests to verify the eventual msg format?
PS2: empty space doesn't look like an appropriate separtor, maybe we should use something like `|` to separate PreFitler/Filter/PostFilter message.

File: pkg/scheduler/schedule_one.go [hunk 3]
```diff
@@ -377,11 +377,9 @@ func (sched *Scheduler) findNodesThatFitPod(ctx context.Context, fwk framework.F
 		if !s.IsUnschedulable() {
 			return nil, diagnosis, s.AsError()
 		}
-		// All nodes will have the same status. Some non trivial refactoring is
-		// needed to avoid this copy.
-		for _, n := range allNodes {
-			diagnosis.NodeToStatusMap[n.Node().Name] = s
-		}
+		// Record the messages from PreFilter in Diagnosis.PreFilterMsg
```
Comment (Huang-Wei) [inline pkg/scheduler/schedule_one.go:380]: nit: Add a period at the end.

File: pkg/scheduler/framework/types.go (hunk 1 above)
Comment (yuanchen8911) [reply pkg/scheduler/framework/types.go]: Made the suggested changes.

File: pkg/scheduler/framework/types.go (hunk 2 above)
Comment (yuanchen8911) [reply pkg/scheduler/framework/types.go]: Updated the format. How about using `.` as the separator. 

`0/X nodes are available: <PreFilterMsg>. <FilterMsg>. <PostFilterMsg>.`

File: pkg/scheduler/schedule_one.go (hunk 3 above)
Comment (yuanchen8911) [reply pkg/scheduler/schedule_one.go]: Done.

**Review by yuanchen8911 — COMMENTED** [review]
@Huang-Wei PTAL. Thanks!

File: pkg/scheduler/framework/types.go [hunk 4]
```diff
@@ -238,6 +241,12 @@ func (f *FitError) Error() string {
 		}
 	}
 
+	reasonMsg := fmt.Sprintf(NoNodeAvailableMsg+":", f.NumAllNodes)
+	// Add the messages from PreFilter plugins to reasonMsg.
+	preFilterMsg := f.Diagnosis.PreFilterMsg
+	if preFilterMsg != "" {
+		reasonMsg += " " + preFilterMsg + "."
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply pkg/scheduler/framework/types.go]: Define and add a constant string for separation of PreFilterMsg, FilterMsg and PostFilterMsg.

File: pkg/scheduler/framework/types.go [hunk 5]
```diff
@@ -230,6 +232,7 @@ type FitError struct {
 const NoNodeAvailableMsg = "0/%v nodes are available"
 
 // Error returns detailed information of why the pod failed to fit on each node
+// A message is formated as "0/X nodes are available: <PreFilterMsg>. <FilterMsg>. <PostFilterMsg>."
```
[the remark this answers is not in the document]
Comment (yuanchen8911) [reply pkg/scheduler/framework/types.go:239]: How about adding Erro() testing  in another PR?

**Comment by yuanchen8911** [comment]
@Huang-Wei Can you review it again? Thanks!

**Review by Huang-Wei — COMMENTED** [review]
Logic LGTM, but it'd be nice to get the changes covered by (integration) tests.

**Comment by yuanchen8911** [comment]
> Logic LGTM, but it'd be nice to get the changes covered by (integration) tests.

@Huang-Wei  Per our discussion offline, I added testing for `FitError.Error()`. Thanks.

**Comment by Huang-Wei** [comment]
> @Huang-Wei Per our discussion offline, I added testing for FitError.Error(). Thanks.

Yup, `TestSchedulerSchedulePod()` is supposed to cover the combination of PreFilter/Filter/PostFilter messages, and in this PR, you cover the correctness of FitError.Error().

/lgtm
/approve

Thanks @yuanchen8911 !

**Pull request merged**